refactor(compiler): simplify parseTemplateBindings() for microsyntax

kyliau opened the pull request:
This commit is purely a refactoring of the logic in
`parseTemplateBindings` method for parsing the microsyntax expression.
This is done to enable the introduction of `keySpan` and `valueSpan` in
subsequent PR.

For a detailed explanation of this work and the subsequent work items,
please see https://docs.google.com/document/d/1mEVF2pSSMSnOloqOPQTYNiAJO0XQxA1H0BZyESASOrE/edit?usp=sharing

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Commit:
refactor(compiler): Break up parseTemplateBindings() for microsyntax

This commit is purely a refactoring of the logic in
`parseTemplateBindings` method for parsing the microsyntax expression.
This is done to enable the introduction of `keySpan` and `valueSpan` in
subsequent PR.

For a detailed explanation of this work and the subsequent work items,
please see https://docs.google.com/document/d/1mEVF2pSSMSnOloqOPQTYNiAJO0XQxA1H0BZyESASOrE/edit?usp=sharing

AndrewKushnir commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -118,7 +118,20 @@ export class Parser {
 
-  parseTemplateBindings(tplKey: string, tplValue: string, location: any, absoluteOffset: number):
+  /**
+   * Parse microsyntax template expression and return errors, if any.
```
May be rephrase this a bit to say something like:
```
Parse microsyntax template expression and return either an AST that represents this expression or parsing errors in case given expression in invalid.
```

AndrewKushnir commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -118,7 +118,20 @@ export class Parser {
+   *
+   * @param tplKey name of directive, without the * prefix. For example: ngIf, ngFor
+   * @param tplValue string value for the microsyntax expression
+   * @param location
```
Can we add a quick description of what this parameter is (to be consistent with other params)?

AndrewKushnir commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -723,77 +736,168 @@ export class _ParseAST {
+      absoluteOffset: number): TemplateBinding[] {
+    const bindings: TemplateBinding[] = [];
+    let key = templateKey;
+    if (rawKey !== templateKey) {
```
Just wondering in which case it would not be the case (may be add a comment on this)?

AndrewKushnir commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -723,77 +736,168 @@ export class _ParseAST {
+    // The binding could optionally be followed by "as". For example,
+    // *ngIf="cond | pipe as x". In this case, the key in the "as" binding
+    // is "x" and the value is the template key itself ("ngIf").
+    const asBinding = this.parseValueAsKey(key, rawKeySpan, absoluteOffset);
```
I was a bit confused by the function name (it sounds like we parse a value as a key) :) May be we can rename it to be more clear on what "as" stands for in function name, like `parseValueOfAsExpression`?

AndrewKushnir commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -723,77 +736,168 @@ export class _ParseAST {
+    if (asBinding) {
+      bindings.push(asBinding);
+    }
+    if (!this.optionalCharacter(chars.$SEMICOLON)) {
```
I think this if statement might benefit from additional comments.

ayazhafiz replied:
+1 and consider renaming to consumeOptionalCharacter

kyliau replied:
Will rename this in standalone PR

AndrewKushnir replied:
@kyliau could you plz add a quick comment on why this `if` statement is here and what's the goal? (I agree that renaming `optionalCharacter` can be a part of a separate PR).

kyliau replied:
Instead of a comment, I refactored this into a standalone method called `consumeStatementTerminator()` and added documentation there.

AndrewKushnir commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -723,77 +736,168 @@ export class _ParseAST {
+   * In the first binding, `item` is bound to `NgForOfContext.$implicit`.
+   * In the second binding, `i` is bound to `NgForOfContext.index`.
+   */
+  private parseLetDeclaration(): TemplateBinding|undefined {
```
I glanced over other functions in the `parser.ts` script and it looks like we typically return `null` when we hit exit condition without any work performed. I think we should follow this in the newly added functions as well.

ayazhafiz replied:
+1

kyliau replied:
done

AndrewKushnir reviewed (changes requested):
Thanks a lot for separating this change out in a standalone PR @kyliau!

FYI I've added a few comments, nothing major, but may worth looking into.

Thank you.

ayazhafiz commented on packages/compiler/src/expression_parser/parser.ts:
````diff
@@ -118,7 +118,20 @@ export class Parser {
+   * For example,
+   * ```
+   *   <div *ngFor="let item of items">
+   *                ^ absoluteOffset for `tplValue`
````
Consider mentioning that absoluteOffset is the start of the template binding relative to the start of the source file it's in. This example doesn't make it clear what absoluteOffset actually is.

ayazhafiz replied:
No, there is no other meaning. I just mean that the comment is non-obvious in that it doesn't tell us much about why the absoluteOffset is noted in the example and why it would be useful. I think it would be helpful to add this information to the comment. But this is a nit.

kyliau replied:
ah I see what you mean. `absoluteOffset` is actually an input parameter to this function, which is why I called it out in the example. I've quoted it in backticks to make this more obvious. Let me know if this is better.
(in the next PR I'm going to differentiate between absoluteKeyOffset and absoluteValueOffset)

ayazhafiz commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -723,77 +736,168 @@ export class _ParseAST {
   /**
    * An identifier, a keyword, a string with an optional `-` in between.
```
```suggestion
   * Parses an identifier, a keyword, a string with an optional `-` in between.
```

ayazhafiz commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -723,77 +736,168 @@ export class _ParseAST {
 
-  // Parses the AST for `<some-tag *tplKey=AST>`
+  /**
+   * Parses the bindings for microsyntax `<some-tag *tplKey=AST>`
```
```suggestion
   * Parses the bindings for a microsyntax expression; e.g. the AST in <some-tag *tplKey=AST>
```

ayazhafiz commented on packages/compiler/src/expression_parser/parser.ts:
````diff
@@ -723,77 +736,168 @@ export class _ParseAST {
+   * ```
+   *   <div *ngFor="let item of items; index as i; trackBy: func">
+   *                ^ absoluteOffset for `this.input`
+   * ```
````
Maybe add an example of what we expect to be parsed out of here? (let items, of expression, "as i" expression, trackBy key expression)

ayazhafiz commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -723,77 +736,168 @@ export class _ParseAST {
+      if (letBinding) {
+        bindings.push(letBinding);
+      } else {
+        const {key, keySpan} = this.expectTemplateBindingKey();
```
Why do we expect a template binding key if this is not a let expression?

ayazhafiz commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -723,77 +736,168 @@ export class _ParseAST {
+    if (rawKey !== templateKey) {
+      // Transform rawKey to actual key. Eg. of -> ngForOf, trackBy -> ngForTrackBy
+      key = templateKey + rawKey[0].toUpperCase() + rawKey.substring(1);
+      this.optionalCharacter(chars.$COLON);  // trackBy: trackByFunction
```
We should probably rename this to consumeOptionalCharacter or similar

ayazhafiz commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -723,77 +736,168 @@ export class _ParseAST {
+  }
+
+  private parseValueAsKey(value: string, valueSpan: ParseSpan, absoluteOffset: number):
+      TemplateBinding|undefined {
```
nit for consistency

```suggestion
      TemplateBinding|null {
```

kyliau commented:
I've addressed all comments, PTAL.
After an offline discussion with @AndrewKushnir, I went back to look at the existing test suites for the expression parser, and found that while there are many test cases, all of them are "fictional", so I added some test cases that better represent real world usage of the two most common microsyntax directives - `ngIf` and `ngFor`.
Please let me know what you think, thank you!

AndrewKushnir commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -118,11 +118,32 @@ export class Parser {
+   *
+   * @param tplKey name of directive, without the * prefix. For example: ngIf, ngFor
+   * @param tplValue RHS of the microsyntax attribute
+   * @param templateUrl template filename if it's external, component filename if it's inline
```
nit: I assume `tpl` in `tplKey` and `tplValue` is a short version of `template`. If that's the case, the `templateUrl` (with full `template` word) seems a bit inconsistent. I think `location` (that you had before) was fine, it just needed a `@param` description.

kyliau replied:
I prefer `templateUrl` over `location`, as the former is less ambiguous. To keep everything consistent I renamed `tplKey` and `tplValue` to `templateKey` and `templateValue`, respectively.

AndrewKushnir commented on packages/compiler/src/expression_parser/parser.ts:
````diff
@@ -118,11 +118,32 @@ export class Parser {
+   *   <div *ngFor="let item of items">
+   *                ^ absoluteOffset for `tplValue`
+   * ```
+   * contains three bindings:
````
May be we can also include the de-sugared version of the same example (de-sugar the `*ngFor`), so it'd be clear why we have 3 bindings?

kyliau replied:
done

AndrewKushnir commented on packages/compiler/test/expression_parser/parser_spec.ts:
```diff
@@ -317,6 +325,44 @@ describe('parser', () => {
       expect(bindings[0].expression !.location).toEqual('location');
     });
 
+    it('should support common usage of ngIf', () => {
```
Thanks for adding extra tests! 👍

AndrewKushnir commented on packages/compiler/test/expression_parser/parser_spec.ts:
```diff
@@ -270,6 +270,14 @@ describe('parser', () => {
+    function humanize(bindings: TemplateBinding[]): Array<[string, string | null, boolean]> {
+      return bindings.map(binding => {
+        const {key, expression, name, keyIsVar} = binding;
+        const value = keyIsVar ? name : expression ? expression.source : expression;
```
May be wrap inner ternary expression into `()`, so that it's easier to read?

kyliau replied:
done. Actually, I'd like to use optional chaining and nullish coalescing here, but our version of clang-format doesn't support this yet. I filed a bug with dev infra.

kyliau reviewed (commented):
Addressed Andrew's comments

AndrewKushnir commented on packages/compiler/src/expression_parser/parser.ts:
````diff
@@ -721,79 +747,199 @@ export class _ParseAST {
+   *   <div *ngFor="let item of items; index as i; trackBy: func">
+   * ```
+   * contains five bindings:
+   * 1. ngFor -> null
````
May be add de-sugared version here as well?

ayazhafiz replied:
Also, consider adding a link to the template syntax grammar (https://angular.io/guide/structural-directives#grammar or https://gist.github.com/mhevery/d3530294cff2e4a1b3fe15ff75d08855). It may not be immediately clear why the binding for ngFor is null.

kyliau replied:
added the second link

AndrewKushnir reviewed (approved):
Thanks for addressing all the comments @kyliau and once again thanks for additional documentation and tests 👍 The changes LGTM, but as we discussed offline, it'd be great if @alxhub would have a look at this PR too, since this PR touches pretty fundamental expression parsing logic (adding Alex as a reviewer to this PR). Thank you.

ayazhafiz commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -721,79 +747,199 @@ export class _ParseAST {
+      if (letBinding) {
+        bindings.push(letBinding);
+      } else {
+        // Two possible cases here, either "value as key" or
```
nit: 
```suggestion
        // Two possible cases here, either `value "as" key` or
```

to make more clear that "as" is a literal rather than a variable. If you choose to apply this, the next two lines should be updated too.

ayazhafiz commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -721,79 +747,199 @@ export class _ParseAST {
         } else {
-          name = '\$implicit';
+          // Otherwise the key must be a directive keyword, like "of". Transform
+          // the key to actual key. Eg. of -> ngForOf, trackBy -> ngForTrackBy
```
nice comment!

ayazhafiz commented on packages/compiler/src/expression_parser/parser.ts:
````diff
@@ -721,79 +747,199 @@ export class _ParseAST {
+   * *ngIf="cond | pipe as foo"
+   *  ^^^^                 ^^^
+   *  value                key
+   * ```
````
Is this always the case? For example, in the branch on line 801 - 810, it looks like the value could be "cond | pipe". Maybe there is a better way to name this function, like `parseAsBinding`?

kyliau replied:
No, the value cannot be "cond | pipe". In an "as" binding **that follows an expression**, the value is always the template key (ngFor, ngForOf, etc). This is documented in `parseDirectiveKeyword`. I've updated the example here to reflect the case in line 801-810.

kyliau replied:
I've renamed
1. `parseDirectiveKeyword` -> `parseDirectiveKeywordBindings`
2. `parseLetDeclaration` -> `parseLetBinding`
3. `parseValueAsKey` -> `parseAsBinding`

Note that (1) is in plural form to reflect the fact that there could be two bindings.

AndrewKushnir commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -118,12 +118,38 @@ export class Parser {
+  /**
+   * Parse microsyntax template expression and return a list of bindings or
+   * parsing errors in case the given expression is invalid.
+
```
Missing `*` char.

AndrewKushnir commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -721,79 +747,203 @@ export class _ParseAST {
+  /**
+   * Parse microsyntax template expression and return a list of bindings or
+   * parsing errors in case the given expression is invalid.
+
```
Missing `*` char.

AndrewKushnir commented on packages/compiler/src/expression_parser/parser.ts:
````diff
@@ -721,79 +747,203 @@ export class _ParseAST {
+   *                  ^^ ^^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^
+   *             keyword    bound target   optional 'as' binding
+   * ```
+
````
Missing `*` char.

alxhub reviewed (approved):
Fantastic work!

kyliau commented:
g3 presubmit: http://test/OCL:298480280:BASE:298966468:1583367302889:60602a0a
